### ISSUE TITLE
allow option for type narrowing in subclass overrides in `reportIncompatibleVariableOverride` and `reportIncompatibleMethodOverride`

### ISSUE THREAD

**Issue by NorthIsUp** [microsoft/pyright#10059]
this proposal is asking for a setting that would allow a subclass to narrow a type from its superclass.

Currently this is the behavior of `reportIncompatibleVariableOverride`
```py
class Parent:
    x: int | str | None
    y: int

class Child(Parent):
    x: int  # Type error: 'x' cannot be redeclared with subtype because variable is mutable and therefore invariant
    y: str  # Type error: 'y' cannot be redeclared with an incompatible type
```

the request is to have an option that allows type narrowing behavior for `x` but not the type incompatible behavior for `y`
```py
class Parent:
    x: int | str | None
    y: int

class Child(Parent):
    x: int  # <-- add option to allow this
    y: str  # Type error: 'y' cannot be redeclared with an incompatible type
```

## motivation

the motivation is to remove unnecessary validation from inside methods that use the type when dealing with a narrowed subclass. 

Using the current behavior, we have to do this:
```py
from pydantic import BaseModel

class Parent(BaseModel):
     # x is optional in the parent class
    x: int | None = None
    y: int
    
class Child(Parent):
    # x is not optional in the child class, it will always be an int
    x: int | None
    y: str  # Type error: 'y' cannot be redeclared with an incompatible type

    def do_something_with_x(self):
        # this satisfies the type checker with a redundant if statement
        # pydantic has already validated this upon construction
        if self.x is None:
            raise ValueError("x is None")
        return self.x + 1

    def do_something_else_with_x(self):
        # this is bad because it is not apparent why we can assert this
        assert self.x is not None
        return self.x + 1

    def do_something_matchy_with_x(self):
        # this is some nice modern syntax!
        # yet it is still unnecessary validation and it is also expensive to execute
        match self.x:
            case None:
                raise ValueError("x is None")
            case int():
                return self.x + 1
```

```py
from pydantic import BaseModel

class Parent(BaseModel):
    # x is optional in the parent class
    x: int | None = None
    y: int


class Child(Parent):
    # x is not optional in the child class
    x: int  # <-- add option to allow this
    y: str  # Type error: 'y' cannot be redeclared with an incompatible type

    def do_something_with_x(self):
        # no additional syntax is needed! yay!
        # this is pythonic and it is also the most performant way to do this
        return self.x + 1
```

**Comment by erictraut** [comment]
While this would eliminate the need for validation, it would also be unsafe from a type perspective. Since the purpose of a type checker is to validate type rules, I'm not convinced that it would make sense to allow such a fundamental check to be disabled.

Also, if we were to ever implement such a feature, it would need to be done by creating two new checks and splitting them out from the existing `reportIncompatibleVariableOverride` and `reportIncompatibleMethodOverride` checks. We have done such splits in the past, but they can be painful for existing pyright users because they have backward compatibility implications. The bar is therefore _very_ high for justifying such a change. 

I'm therefore going to reject this enhancement request.

Now, let's consider other potential solutions for your use case that don't involve disabling fundamental type checks.

Covariant overrides are safe in cases where the base class attribute is read-only. There is a new PEP under development ([PEP 767](https://peps.python.org/pep-0767/)) that would allow the `ReadOnly` type qualifier to be used outside of `TypedDict` definitions. That might address your use case.

Another approach that can retain type safety but allow covariant overrides is through the use of generics. You may be able to employ some variant of this for your use case.

```python
class Parent[T: int | None]:
    x: T

class Child(Parent[int]):
    x: int
```

**Comment by NorthIsUp** [comment]
Thank you for the thoughtful response. The generics might work for my specific use case but its worth noting the ergonomics wouldn't scale for a large number of fields.

The concern about the heavy lift/backwards incompatibility is understandable.

but I do want to push on the "unsafe from a type perspective" issue. (Apologies if this is take is naive, I don't know the pyright internals at all.)

> While this would eliminate the need for validation, it would also be unsafe from a type perspective. Since the purpose of a type checker is to validate type rules, I'm not convinced that it would make sense to allow such a fundamental check to be disabled.

Does this proposal actually contradict this? Having a subclass with a field that is a strict subset of the types of the field in the parent class shouldn't be unsafe. The type checker should see that the type has been narrowed and then use that information when validating type rules. e.g. if the type is `str | None` and the subclass type is `str` then `str` will _always_ satisfy `str | None`.

again, thanks for the time and the excellent work y'all are doing on this project 🙏

**Comment by NorthIsUp** [comment]
ah, thank you! makes sense now

**Comment by NorthIsUp** [comment]
#funfact I worked at Skype before it was acquired!
